twill-yarn/src/main/java/org/apache/twill/yarn/YarnTwillPreparer.java

```diff
@@ -384,6 +408,24 @@ private void setEnv(String runnableName, Map<String, String> env, boolean overwr
     }
   }
 
+  private void saveLogLevels(LogEntry.Level level) {
+    level = level == null ? LogEntry.Level.INFO : level;
+    Map<String, String> appLogLevels = new HashMap<>();
+    appLogLevels.put(Logger.ROOT_LOGGER_NAME, level.name());
+    for (String runnableName : twillSpec.getRunnables().keySet()) {
+      this.logLevels.put(runnableName, appLogLevels);
+    }
+  }
+
+  private void saveLogLevels(String runnableName, Map<String, LogEntry.Level> logLevels) {
+    Map<String, String> newLevels = new HashMap<>();
+    for (Map.Entry<String, LogEntry.Level> entry : logLevels.entrySet()) {
+      Preconditions.checkArgument(entry.getValue() != null, "Log level cannot be null for logger {}", entry.getKey());
+      newLevels.put(entry.getKey(), entry.getValue().name());
+    }
+    this.logLevels.get(runnableName).putAll(newLevels);
```

Comment:
What I meant was, since the method is "set", it should be setting / overwriting, not appending. And I think it's ok for the user to gather and pass in the map.